the first presentation

AlexanderSparrow opened the pull request.

AlexanderSparrow committed:
Delete .idea directory

nothing superfluous

AlexanderSparrow committed:
Delete .github/workflows directory

nothing superfluous

AlexanderSparrow committed:
Delete java-kanban.iml

nothing superfluous

AlexanderSparrow committed:
Delete .gitignore

nothing superfluous

AlexanderSparrow committed:
Merge branch 'refs/heads/sprint_7-solution-in-file-manager'

# Conflicts:
#	.gitignore

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 1
а the first presentation

kesch9 commented on src/model/Epic.java:
```diff
@@ -2,35 +2,94 @@
 public class Epic extends Task {
     private final Set<SubTask> subTasks;
+    private Duration duration;
+    private LocalDateTime startTime;
```
Обязательно
duration и startTime наследуются от Task не нужно их писать ещё раз, не забываем про один из принципов ООП

AlexanderSparrow replied:
Done

kesch9 replied:
Неверно подправили, endTime не нужно было удалять нужно вернуть. В комментарии написано было только про duration и startTime, они есть в Task, а endTime нет.

AlexanderSparrow replied:
задал поле endTime

kesch9 commented on src/model/Epic.java:
```diff
@@ -2,35 +2,94 @@
+        this.endTime = null;
+        this.duration = Duration.ZERO;
+        //updateEpicStatus();
+        calculateFields();
```
Обязательно
Не нужно вызывать calculateFields при создании эпика, вы и так явно их обнуляете

AlexanderSparrow replied:
Done

kesch9 replied:
Неверно подправлено, удалять конструктор с новыми полями не нужно было, этот конструктор должен использоваться при восстановлении

AlexanderSparrow replied:
Вернул конструктор с новыми полями

kesch9 commented on src/model/Epic.java:
```diff
@@ -2,35 +2,94 @@
+        this.startTime = null;
+        this.endTime = null;
+        this.duration = Duration.ZERO;
+        //updateEpicStatus();
```
Обязательно
Закомментированного кода не должно быть на ревью

AlexanderSparrow replied:
Done

AlexanderSparrow replied:
Done

kesch9 replied:
Не подправлено, в 83 строчке опять закоментированный код

AlexanderSparrow replied:
Удалил закоментированный код в 83 строчке

kesch9 commented on src/model/Epic.java:
```diff
@@ -2,35 +2,94 @@
+        this.endTime = null;
+        this.duration = Duration.ZERO;
+        //updateEpicStatus();
+        calculateFields();
```
Обязательно
Не нужно вызывать calculateFields при создании эпика, вы и так явно их обнуляете

AlexanderSparrow replied:
Done

kesch9 commented on src/model/Epic.java:
```diff
@@ -2,35 +2,94 @@
+        this.startTime = null;
+        this.endTime = null;
+        this.duration = Duration.ZERO;
+        //updateEpicStatus();
```
Обязательно
Закомментированного кода не должно быть на ревью

AlexanderSparrow replied:
Done

AlexanderSparrow replied:
Done

kesch9 replied:
Не подправлено, в 83 строчке опять закоментированный код

AlexanderSparrow replied:
Удалил закоментированный код в 83 строчке

kesch9 commented on src/service/CsvTaskParser.java:
```diff
@@ -2,61 +2,94 @@
+                return new Task(id, name, description, status, duration, startTime);
+            case EPIC:
+                LocalDateTime endTime = fields.length > 7 ? LocalDateTime.parse(fields[7], DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")) : null;
+                return new Epic(id, name, description, duration, startTime, endTime);
```
Обязательно
В конструкторе эпика вы сбрасываете duration, startTime, endTime, поэтому вы ничего не восстанавливаете

AlexanderSparrow replied:
Done

kesch9 replied:
Вы все еще не восстанавливаете endTime т.к. удалили поле.  И удалять из конструктора новые поля не нужно было.

AlexanderSparrow replied:
Восстанавливаем все поля у Epic

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -7,33 +7,157 @@
 
     public FileBackedTaskService(File file) {
         this.file = file;
+        this.prioritizedTasks = new TreeSet<>(new TaskStartTimeComparator());
```
Обязательно
Сортированный список и все проверки на пересечения должны быть в InMemoryTaskService, а не  FileBackedTaskService.

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -7,33 +7,157 @@
+                        }
+                    });
+
+            getEpics().stream()
```
Обязательно
Здесь нужно запрашивать именно подзадачи, а не эпики, чтобы затем делать flatMap. Просто получаем все подзадачи и пишем их

AlexanderSparrow replied:
Done

kesch9 replied:
Не подправлено

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -7,33 +7,157 @@
         }
     }
 
+//    private void save() {
```
Обязательно
На ревью не должно быть закомментированного кода, плюс у вас есть гит в котором сохраняется все изменения и вы можете их посмотреть

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -7,33 +7,157 @@
+        }
+        return fileBackedTaskService;
+    }
+    /*public static FileBackedTaskService loadFromFile(File file) {
```
Обязательно
На ревью не должно быть закомментированного кода, плюс у вас есть гит в котором сохраняется все изменения и вы можете их посмотреть

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -7,33 +7,157 @@
+                        Epic epic = (Epic) CsvTaskParser.fromCsvString(line);
+                        assert epic != null;
+                        fileBackedTaskService.epics.put(epic.getId(), epic);
+                        prioritizedTasks.add(epic);
```
Обязательно
Эпики не храняться в сортированном списке

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     public void addEpic(Epic epic) {
+        // Эпики обычно не имеют времени выполнения, поэтому проверка не требуется
         super.addEpic(epic);
+        prioritizedTasks.add(epic);
```
Обязательно
Эпики не хранятся в сортированном списке

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
 
     @Override
     public void updateEpic(Epic epic) {
+        prioritizedTasks.remove(getEpicById(epic.getId())); // Удаляем старую версию эпика
```
Обязательно
Эпики не хранятся в сортированном списке

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     @Override
     public void removeAllEpics() {
         super.removeAllEpics();
+        prioritizedTasks.clear();
```
Обязательно
Эпики не хранятся в сортированном списке, но при удалении всех эпиков нужно удалить подзадачи из сортированного списка

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     }
 
-    public static void main(String[] args) {
+    /*public static void main(String[] args) {
```
Обязательно
На ревью не должно быть закомментированного кода, плюс у вас есть гит в котором сохраняется все изменения и вы можете их посмотреть

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
 
     @Override
     public void removeTask(int id) {
+        Task task = getTaskById(id);
```
Обязательно
Подзадачу получаем напрямую из хранилища, не используем метод getTaskById иначе добавиться в историю, этого не нужно делать

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     @Override
     public void removeTask(int id) {
+        Task task = getTaskById(id);
+        prioritizedTasks.remove(task);
```
Обязательно
Если такой задачи нет в хранилище по id, то и удалять не нужно, нам вернется null

AlexanderSparrow replied:
Done

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
 
     @Override
     public void removeSubTask(int id) {
+        SubTask subTask = getSubTaskById(id);
```
Обязательно
Подзадачу получаем напрямую из хранилища, не используем метод getTaskById иначе добавиться в историю, этого не нужно делать

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     @Override
     public void removeSubTask(int id) {
+        SubTask subTask = getSubTaskById(id);
+        prioritizedTasks.remove(subTask);
```
Обязательно
Если такой задачи нет в хранилище по id, то и удалять не нужно, нам вернется null

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     @Override
     public void removeEpic(int id) {
+        Epic epic = getEpicById(id);
+        prioritizedTasks.remove(epic);
```
Обязательно
Эпики не хранятся в сортированном списке, но при удалении эпика нужно удалить его подзадачи из сортированного списка

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
 
     @Override
     public void updateTask(Task task) {
+        if (hasTimeOverlap(task)) {
```
Обязательно
При проверки на пересечения нужно исключить при анализе саму задачу, в stream отфильтровать по id

kesch9 replied:
Также при обновлении нужна проверка, что эта задача есть в хранилище

AlexanderSparrow replied:
Done

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
+        if (hasTimeOverlap(task)) {
+            throw new IllegalArgumentException("Задача пересекается по времени с другой задачей.");
+        }
+        prioritizedTasks.remove(getTaskById(task.getId())); // Удаляем старую версию задачи
```
Обязательно
Подзадачу получаем напрямую из хранилища, не используем метод getTaskById иначе добавиться в историю, этого не нужно делать

AlexanderSparrow replied:
Done(?)

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -75,83 +207,111 @@ public static FileBackedTaskService loadFromFile(File file) {
     @Override
     public void removeAllTasks() {
         super.removeAllTasks();
+        prioritizedTasks.clear();
```
Обязательно
Удаляются только задачи, а не чиститься весь сортированный список, т.к. там хранятся еще подзадачи

AlexanderSparrow replied:
Done

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -69,6 +69,7 @@ public void addSubTask(SubTask subTask) {
         subTasks.put(id, subTask);
         subTask.getEpic().getSubTasks().add(subTask);
         subTask.getEpic().updateEpicStatus();
+        subTask.getEpic().calculateFields();
```
Обязательно
Вы добавили пересчет времени у эпик только при добавлении, но нужно как и статус пересчитывать  при удалении и обновлении подзадач. Также вы три раза получаете подзадачу из хранилища, что не эффективно. Получите подзадачу и сохраните в локальную переменную и используйте её. Представьте если вы получаете по сети данные или считываете из файла, вы три раза будете это делать, что крайне плохо скажется на производительности

AlexanderSparrow replied:
использовал локальную переменную для хранения подзадачи

AlexanderSparrow replied:
методы removeSubTask и addSubTask поправил по аналогии

kesch9 commented on src/service/testUpdateEpicStatusAllNew.java:
```diff
@@ -0,0 +1,4 @@
+package service;
+
+public class testUpdateEpicStatusAllNew {
```
Обязательно
Пустых классов не должно быть

AlexanderSparrow replied:
Done

kesch9 commented on test/FileBackedTaskServiceTest.java:
```diff
@@ -1,19 +1,26 @@
 import static org.junit.jupiter.api.Assertions.*;
 
-class FileBackedTaskManagerTest {
+class FileBackedTaskServiceTest {
```
Обязательно
Не выполнено требование
_Для двух менеджеров задач InMemoryTaskManager и FileBackedTaskManager.
a. Чтобы избежать дублирования кода, нужен базовый класс с тестами на каждый метод из интерфейса abstract class TaskManagerTest<T extends TaskManager>.
b. Для подзадач нужно дополнительно проверить наличие эпика, а для эпика — расчёт статуса.
c. Добавить тест на корректность расчёта пересечения интервалов._

AlexanderSparrow replied:
расчет статуса проверяется в EpicTest 
проверка корректности эпика у подзадачи проверяетяс в SubTaskTest
проверка корректности персечения интервалов добавлена

kesch9 commented on test/FileBackedTaskServiceTest.java:
```diff
@@ -64,22 +75,24 @@ void shouldSaveAndLoadTasksCorrectly() {
         assertEquals(subTask1.getDescription(), loadedSubTask1.getDescription(), "Ошибка: Описание SubTask не совпадает");
         assertEquals(subTask1.getStatus(), loadedSubTask1.getStatus(), "Ошибка: Статус SubTask не совпадает");
         assertEquals(subTask1.getEpicId(), loadedSubTask1.getEpicId(), "Ошибка: Epic ID у SubTask не совпадает");
+        assertEquals(subTask1.getId(), loadedSubTask1.getId(), "Ошибка: ID SubTask не совпадает");
```
Обязательно
Нужно в тест добавить восстановления новых полей

AlexanderSparrow replied:
Выполнено

kesch9 reviewed (commented):
Нужно многое подправить в работе

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 1
а the second presentation

kesch9 commented on src/model/Task.java:
```diff
@@ -61,6 +75,30 @@ public void setId(int id) {
+        this.startTime = startTime;
+    }
+    public void setEndTime(LocalDateTime startTime) {
+        this.startTime = startTime;
```
Обязательно
setEndTime должен быть в эпике, здесь вообще реализовано неверно, метод называется endTime, а задаете startTime, т.е. вы затираете неправильным значением начало времени

AlexanderSparrow replied:
Метод перенесен в Epic

kesch9 commented on src/model/Task.java:
```diff
@@ -1,18 +1,24 @@
     protected String description;
     protected Status status;
+    private Duration duration;
+    private LocalDateTime startTime;
```
Обязательно
Новые поля protected

AlexanderSparrow replied:
Новые поля теперь protected

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -12,7 +13,9 @@ public class InMemoryTaskService implements TaskService {
-    private final HistoryService historyService = Services.getDefaultHistory();
+    protected final TreeSet<Task> prioritizedTasks = new TreeSet<>(new TaskStartTimeComparator());
+
+    final HistoryService historyService = Services.getDefaultHistory();
```
Обязательно
Для historyService добавляем protected

AlexanderSparrow replied:
Для historyService добавил protected

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -133,75 +107,52 @@ public void updateEpic(Epic epic) {
+                    if (task instanceof SubTask subTask) {
+                        Epic epic = service.getEpicById(subTask.getEpicId());
+                        task = CsvTaskParser.fromCsvString(line, epic);
+                        service.addSubTask(subTask);
```
Обязательно
Верните восстановление как было, вы сделали неверно, НЕЛЬЗЯ добавлять через метод addSubTask, НЕЛЬЗЯ вызывать метод getEpicById, т.к. будет записываться в файл и в историю. Мы в прошлой итерации столько времени потратили на исправления и вы сделали опять неправильно. Во вторых даже идея подсвечиывает 140 строчку, что вы task не используете, а запиваете subTask, что неверно. Вы пишите тип задачи, прочитайте эту строчку и сделайте проверку type == "Task" и т.д., не нужно сначала восстанавливать задачу, затем еще раз восстанавливать задачу. Выяснили какой тип и только после этого восстановить нужный тип задачи

AlexanderSparrow replied:
Исправлено

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -133,75 +107,52 @@ public void updateEpic(Epic epic) {
+                    } else if (task instanceof Task) {
+                        service.addTask(task);
+                    }
+                } catch (IllegalArgumentException e) {
```
Обязательно
Unchecked ошибку не нужно ловить или ловите вместо IOException все Exception

AlexanderSparrow replied:
выполнено

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -133,75 +107,52 @@ public void updateEpic(Epic epic) {
+        if (!file.exists()) {
+            try {
+                file.getParentFile().mkdirs(); // Создаем директорию, если она не существует
+                file.createNewFile(); // Создаем файл
```
Обязательно
Ничего создавать не нужно при восстановлении

AlexanderSparrow replied:
Удален блок кода, отвечающий за создание файлов и директорий, если они не существуют.

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -49,13 +52,43 @@ public Set<SubTask> getAllSubTasksByEpic(int id) {
+            tasks.put(id, task);
+            prioritizedTasks.add(task);
+
+        } catch (IllegalArgumentException e) {
```
Обязательно
Ну нужно ловить ошибки, которые вы сами кидаете

AlexanderSparrow replied:
Убрал, поправил тест

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -101,51 +141,77 @@ public Epic getEpicById(int id) {
+            if (epic != null) {
+                epic.getSubTasks().add(updatedSubTask);
+                epic.updateEpicStatus(); // Обновляем статус эпика
+                epic.calculateFields();// Обновляем поля эпика
```
Обязательно
Вы делаете обновление два раза, сделайте в одном месте обновление, сначала удалите подзадачу, затем добавьте новую и обновите поля.

AlexanderSparrow replied:
Убрал задвоение операций обновления статуса и полей

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -49,13 +52,43 @@ public Set<SubTask> getAllSubTasksByEpic(int id) {
+
+    private boolean hasTimeOverlap(Task newTask) {
+        return prioritizedTasks.stream()
+                .anyMatch(existingTask -> isTimeOverlap(newTask, existingTask));
```
Обязательно
В прошлом ревью я написал, что нужно исключить саму задачу из алгоритма поиска пересечения, иначе при обновлении если время не поменялось, то будет сама с собой конфликтовать. Если у вас не конфликтует, значит алгоритм не верный

AlexanderSparrow replied:
Исключил саму задачу из алгоритма поиска пересечения

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -164,6 +230,7 @@ public void updateEpic(Epic epic) {
     @Override
     public void removeAllTasks() {
         tasks.clear();
+        prioritizedTasks.removeIf(task -> !(task instanceof SubTask || task instanceof Epic)); // Удаляем только задачи
```
Обязательно
Вы удаляете только задачи, не нужно проверять на отрицание,
prioritizedTasks.removeIf(task -> task instanceof Task);

AlexanderSparrow replied:
Проверка на отрицание убрана

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -187,5 +255,25 @@ public void removeAllEpics() {
         }
         subTasks.clear();
         epics.clear();
+        prioritizedTasks.clear();
```
Обязательно
Удаляются только подзадачи

AlexanderSparrow replied:
Удалил только подзадачи

kesch9 commented on src/model/Epic.java:
```diff
@@ -39,4 +75,23 @@ public void updateEpicStatus() {
+        if (subTasks == null || subTasks.isEmpty()) {
+            this.setDuration(Duration.ZERO);
+            this.setStartTime(null);
+            //subTasks = null;
```
Обязательно
Нужно обнулить еще endTime

AlexanderSparrow replied:
endTime обнулен, закоментированный код удален

kesch9 commented on test/FileBackedTaskServiceTest.java:
```diff
@@ -0,0 +1,23 @@
+import java.io.File;
+import java.io.IOException;
+
+public class FileBackedTaskServiceTest extends TaskServiceTest {
```
Обязательно
FileBackedTaskService наследуется от InMemoryTaskService, поэтому должны проходить все тесты для TaskService, но в нем есть еще сохранение и восстановления и их надо отдельно проверить. Не нужно было удалять тесты на сохранения и восстанавления. Их нужно вернуть и дополнительно дополнить проверкой восстановления новых полей с сортированном списом. В прошлом ревью об этом было сказано

AlexanderSparrow replied:
Выполнено

kesch9 commented on test/TaskServiceTest.java:
```diff
@@ -0,0 +1,186 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+        public abstract  class TaskServiceTest {
```
Обязательно
Не верное форматирования файла, лишние отступы в начале

AlexanderSparrow replied:
форматирование выполнено

kesch9 reviewed (commented):
Исправлено неверно

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 3
а the third presentation

AlexanderSparrow reviewed (commented):
Замечания устранены

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 3
а the third presentation (fix some bags)

kesch9 commented on src/service/CsvTaskParser.java:
```diff
@@ -2,61 +2,94 @@
+            case TASK:
+                return new Task(id, name, description, status, duration, startTime);
+            case EPIC:
+                return new Epic(id, name, description, duration, startTime, endTime);
```
Обязательно
У эпика так и не восстанавливаются новые поля, т.к. в конструкторе вы их обнуляете, и в тестах нет проверки этих полей

AlexanderSparrow replied:
В конструкторе, полям присваиваются извлеченные из файла значения

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -133,75 +105,53 @@ public void updateEpic(Epic epic) {
+                            System.err.println("Epic with ID " + epicId + " not found for SubTask.");
+                        } else {
+                            SubTask subTask = CsvTaskParser.fromCsvString(line, epic);
+                            fileBackedTaskService.subTasks.put(subTask.getId(), subTask);
```
Обязательно
Не восстанавливается сортированный список

AlexanderSparrow replied:
Cортированный список восстанавливается

kesch9 commented on test/FileBackedTaskServiceTest.java:
```diff
@@ -64,22 +76,39 @@ void shouldSaveAndLoadTasksCorrectly() {
-    }
 
+        // Проверка сортировки задач в приоритетном списке
+        List<Task> prioritizedTasks = new ArrayList<>(taskService.getPrioritizedTasks());
```
Обязатильно
Нужно проверять, что сортированный список восстановился и запрашивать нужно у loadedManager

AlexanderSparrow replied:
выполнено

kesch9 commented on test/FileBackedTaskServiceTest.java:
```diff
@@ -64,22 +76,39 @@ void shouldSaveAndLoadTasksCorrectly() {
         assertEquals(epic1.getName(), loadedEpic1.getName(), "Ошибка: Имя Epic не совпадает");
         assertEquals(epic1.getDescription(), loadedEpic1.getDescription(), "Ошибка: Описание Epic не совпадает");
         assertEquals(epic1.getStatus(), loadedEpic1.getStatus(), "Ошибка: Статус Epic не совпадает");
+        assertEquals(epic1.getId(), loadedEpic1.getId(), "Ошибка: ID Epic не совпадает");
```
Обязательно
У эпика не проверяется, восстановления новых полей

AlexanderSparrow replied:
Добавлена проверка всех полей эпика при восстановлении

kesch9 commented on src/service/FileBackedTaskService.java:
```diff
@@ -133,75 +105,53 @@ public void updateEpic(Epic epic) {
+    // Метод для загрузки данных из файла
+    public static FileBackedTaskService loadFromFile(File file) {
+        FileBackedTaskService fileBackedTaskService = new FileBackedTaskService(file);
+        AtomicInteger currentMaxId = new AtomicInteger();
```
Применять AtomicInteger излишнее, у вас не будет восстанавливаться в разных потоках, посмотрите для чего нужен AtomicInteger

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -12,7 +13,9 @@ public class InMemoryTaskService implements TaskService {
-    private final HistoryService historyService = Services.getDefaultHistory();
+    protected final TreeSet<Task> prioritizedTasks = new TreeSet<>(new TaskStartTimeComparator());
+
+    final protected HistoryService historyService = Services.getDefaultHistory();
```
Обязательно
Порядок ключевых слов неверный должно быть
protected final

AlexanderSparrow replied:
Установил верный порядок ключевых словв

kesch9 commented on src/model/Epic.java:
```diff
@@ -39,4 +84,28 @@ public void updateEpicStatus() {
+                .map(SubTask::getDuration)
+                .reduce(Duration.ZERO, Duration::plus));
+
+        this.setStartTime(subTasks.stream()
```
Обязательно
Нет пересчета endTime

AlexanderSparrow replied:
Добавлен перерасчет endTime

kesch9 commented on src/model/Epic.java:
```diff
@@ -39,4 +84,28 @@ public void updateEpicStatus() {
+    }
+
+    public void setEndTime(LocalDateTime endTime) {
+        this.startTime = endTime;
```
Обязательно
Написано сеттер endTime вы обновляете startTime

kesch9 replied:
И сеттеры не нужны. Метод setEndTime не нужен, вы можете напрямую писать в поля

kesch9 replied:
Можно удалить сеттер, вы его используете в одном месте напишите вот так
        // Находим максимальное время окончания среди всех подзадач
        endTime = subTasks.stream()
                .map(SubTask::getEndTime)
                .filter(Objects::nonNull)
                .max(LocalDateTime::compareTo)
                .orElse(null);

kesch9 replied:
Вместо
 // Находим максимальное время окончания среди всех подзадач
        this.setEndTime(subTasks.stream()

kesch9 commented on src/model/Epic.java:
```diff
@@ -39,4 +84,28 @@ public void updateEpicStatus() {
+        if (subTasks == null || subTasks.isEmpty()) {
+            this.setDuration(Duration.ZERO);
+            this.setStartTime(null);
+            this.setEndTime(null);
```
Обязательно
Внутри классов не используются сеттеры для своих полей, у вас есть прямой доступ к ним, устанавливаете напрямую

AlexanderSparrow replied:
обращение к полям напрямую выполнено

kesch9 replied:
А в методе calculateFields также через сеттеры

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -164,6 +222,7 @@ public void updateEpic(Epic epic) {
     @Override
     public void removeAllTasks() {
         tasks.clear();
+        prioritizedTasks.removeIf(task -> task instanceof Task); // Удаляем только задачи
```
Не самый лучший вариант написания, даже IDEA подсвечивает этот момент, лучше переписать вот так
tasks.values().forEach(prioritizedTasks::remove);
tasks.clear();

AlexanderSparrow replied:
Исправил

kesch9 commented on src/service/InMemoryTaskService.java:
```diff
@@ -185,7 +245,30 @@ public void removeAllEpics() {
         }
+        for (Epic epic : epics.values()) {
+            // Удаляем все подзадачи эпика из приоритетного списка
+            prioritizedTasks.removeAll(epic.getSubTasks());
```
Вот тоже хороший вариант удаления коллекциями, вместо removeIf

AlexanderSparrow replied:
Спасибо:)

kesch9 reviewed (commented):
Еще есть над чем работать

AlexanderSparrow committed:
PROJECT "Task Tracker". Part Forth: New functionality has been …
…added (task duration and priority)

Task verification iteration: 4
а the fourth presentation (fix some bags)

kesch9 reviewed (approved):
Approve
